statsd/dispatcher.go

```diff
@@ -0,0 +1,185 @@
+package statsd
+
+import (
+	"hash/adler32"
+	"sync"
+	"time"
+
+	"github.com/atlassian/gostatsd/types"
+
+	log "github.com/Sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+// Dispatcher is responsible for managing Aggregators' lifecycle and dispatching metrics among them.
+type Dispatcher interface {
+	Run(context.Context) error
+	DispatchMetric(context.Context, *types.Metric) error
+	Flush(context.Context) <-chan *types.MetricMap
+	Process(context.Context, ProcessFunc) *sync.WaitGroup
+}
+
+// AggregatorFactory creates Aggregator objects.
+type AggregatorFactory interface {
+	// Create creates Aggregator objects.
+	Create() Aggregator
+}
+
+// AggregatorFactoryFunc type is an adapter to allow the use of ordinary functions as AggregatorFactory.
+type AggregatorFactoryFunc func() Aggregator
+
+// Create calls f().
+func (f AggregatorFactoryFunc) Create() Aggregator {
+	return f()
+}
+
+type flushCommand struct {
+	ctx    context.Context
+	wg     sync.WaitGroup
+	result chan<- *types.MetricMap
+}
+
+type processCommand struct {
+	wg sync.WaitGroup
+	f  ProcessFunc
+}
+
+type worker struct {
+	metrics     chan *types.Metric
+	flushChan   chan *flushCommand
+	processChan chan *processCommand
+	aggr        Aggregator
+}
+
+type dispatcher struct {
+	workers map[uint16]*worker
+}
+
+// NewDispatcher creates a new Dispatcher with provided configuration.
+func NewDispatcher(numWorkers int, perWorkerBufferSize int, af AggregatorFactory) Dispatcher {
+	workers := make(map[uint16]*worker)
+
+	n := uint16(numWorkers)
+
+	for i := uint16(0); i < n; i++ {
+		workers[i] = &worker{
+			make(chan *types.Metric, perWorkerBufferSize),
```

Comment:
Makes sense. Apparently gofmt is not enough and people still use different styles for no good reason :)